### PR TITLE
[WIP, DO NOT MERGE] bpo-41188: Prepare CPython for opague PyObject structure.

### PR DESCRIPTION
**Pull request by WildCard65**
The end goal of this pull request is to make the structure "PyObject" opaque to EVERYTHING except Python core.

To accomplish this, the following must be done:
- [x] Add new member to 'PyTypeObject' to hold pointer offset to type's data structure.
- [ ] Add implementation defined "slib_python310" static library for end users to link against.
- [ ] Determine if more changes are required.

https://bugs.python.org/issue41188

### CODE REVIEW

File: Include/object.h [hunk 1]
```diff
@@ -317,6 +317,9 @@ Code can use PyType_HasFeature(type_ob, flag_value) to test whether the
 given type object has a specified feature.
 */
 
+/* Set if the type object's tp_basicsize is set for opague object */
+#define Py_TPFLAGS_OPAQUE_OBJECT (1UL << 8)
```
Comment (vstinner) [inline Include/object.h]: I would prefer a more explicit name, lilke: `Py_TPFLAGS_OMIT_PYOBJECT_SIZE`.

**Review by vstinner — COMMENTED** [review]
> Add implementation defined "slib_python310" static library for end users to link against.

I don't understand this.

> Change as many types as possible to use 'Py_TPFLAGS_USES_OPAQUE_OBJECT'

Please don't do that in the same PR.

**Comment by vstinner** [comment]
I would prefer to open a separated issue on bugs.python.org for this work.

**Comment by WildCard65** [comment]
@vstinner The static library will contain '.c' files that reference internal implementations FOR the runtime it's designed for.
The goal of the static library is that end users link against it when building their extensions WITHOUT having access to internals directly.

As it currently stands, the current Py_TYPE still requires direct access to internals for end users while this static library remove that requirement.

example would be:
Py_TYPE macro references a method called "PyObject_GetType", but the definition of the method is not in the header, instead, it's in the static library which will be linked into the final binary.

This allows for implementation independent access while still allowing the C compiler to aggressively inline out the small functions.

**Comment by WildCard65** [comment]
@vstinner Before I proceed any further, I want you to review "slib_pythoncore" first.

**Comment by vstinner** [comment]
>  The goal of the static library is that end users link against it when building their extensions WITHOUT having access to internals directly.

This is the purpose of the limited C API which provides a stable ABI. It's already implemented since Python 3.2. I don't see the point of adding a second flavor of the "libpython" library.

**Comment by WildCard65** [comment]
As it stands, functions like _Py_TYPE, _Py_SIZE, _Py_INCREF, _Py_DECREF rely on "PyObject" being a complete type. Opaque objects on the other hand are classified as incomplete types.

Incomplete types can't be used for anything but pointers (*) and references (&), these functions will fail when PyObject goes opaque in the limited API, infact, sizeof() also fails on incomplete types.

The purpose of the static library is not to be a different flavour of Python, but to act as a bridge into Python's internals while respecting the limitations of incomplete types.

As an added bonus, this static library can be "reskinned" for runtimes that take a different approach to PyObject (if any) and descendant built-in types while end users don't have to worry about what the runtime's schematic is.

**Comment by WildCard65** [comment]
Added a basic implementation of the flag, but I do not know where all usages of "tp_basicsize" is, I'm also debating on whether or not I need to make GET_TYPE_TOTALSIZE into an inlinable function.

File: Include/cpython/object.h [hunk 2]
```diff
@@ -263,6 +263,10 @@ struct _typeobject {
 
     destructor tp_finalize;
     vectorcallfunc tp_vectorcall;
+
+    /* INTERNAL USE ONLY! MODIFYING THIS CAN CRASH PYTHON! */
+    const Py_ssize_t tp_obj_offset; /* Offset from "PyObject *" pointer */
```
Comment (vstinner) [inline Include/cpython/object.h:268]: I don't think that this is needed. When I get a "PyObject *ob", the offset to the PyObject structure is 0: "PyObject copy = *ob;" in Python internals is valid.

File: Include/cpython/object.h [hunk 3]
```diff
@@ -263,6 +263,10 @@ struct _typeobject {
...
+    const Py_ssize_t tp_obj_size; /* Total memory allocation size */
```
Comment (vstinner) [inline Include/cpython/object.h:269]: I don't see the point of adding a new member. You can just check for Py_TPFLAGS_USES_OPAQUE_OBJECT flag in functions like PyType_GenericAlloc(), _PyObject_SIZE() and _PyObject_VAR_SIZE(). If the flag is set, add sizeof(PyObject).

File: Include/object.h [hunk 4]
```diff
@@ -102,26 +102,62 @@ typedef struct _typeobject PyTypeObject;
...
-
 /* Cast argument to PyVarObject* type. */
 #define _PyVarObject_CAST(op) ((PyVarObject*)(op))
 #define _PyVarObject_CAST_CONST(op) ((const PyVarObject*)(op))
 
+Py_SLIB_LOCAL(Py_ssize_t) PyObject_GetRefCount(const PyObject *ob);
+Py_SLIB_LOCAL(PyTypeObject *) PyObject_GetType(const PyObject *ob);
+
+Py_SLIB_LOCAL(void) PyObject_SetRefCount(PyObject *ob, const Py_ssize_t refcnt);
+Py_SLIB_LOCAL(void) PyObject_SetType(PyObject *ob, const PyTypeObject *type);
```
Comment (vstinner) [inline Include/object.h:133]: Python 3.9 already have Py_SET_REFCNT() and Py_SET_TYPE() functions. What is the advantage of adding new functions?

**Review by vstinner — COMMENTED** [review]
IMHO you gone too far in the implementation before discussing the design. You should first propose the design on python-ideas or python-dev list.

File: Include/cpython/object.h (hunk 2 above)
Comment (WildCard65) [reply Include/cpython/object.h:268]: It's not required if everything inherited ONLY PyObject, that will never be true.

The offset is calculated from the immediate base type of the type, which can be for example: list, tuple, dict, type, custom types.

Same thing applies to "tp_obj_size", without these members, heap corruption can be very likely.

File: Include/cpython/object.h (hunk 3 above)
Comment (WildCard65) [reply Include/cpython/object.h:269]: See previous comment.

File: Include/object.h (hunk 4 above)
Comment (WildCard65) [reply Include/object.h:133]: These aren't essentially "new functions", they are bridges into CPython's internals WITHOUT leaking implementation details.

As mentioned before, the existing functions WILL FAIL when PyObject goes full opaque.

**Comment by vstinner** [comment]
Py_TPFLAGS_USES_OPAQUE_OBJECT looks useful, but other changes like "slib" looks useless to me.

**Review by vstinner — CHANGES_REQUESTED** [review]
I don't see the point of tp_obj_offset, tp_obj_size and slib.

**Comment by WildCard65** [comment]
Consider the following:

```
strict myobject;

void use_myobject(myobect *obj); // Valid, incomplete types can be used as pointers.

size_t get_object_value(my object *obj)
{
    return obj->m_value; // ERROR: my_object is an incomplete type!
}
```

Above demonstrates why functions like _Py_Type will fail in limited api.

'Py_TPFLAGS_OMIT_OBJECT_SIZE' was intended as the stepping stone for making types like PyTypeObject opaque as well, this makes knowing the true offset off the base time a runtime dependant scenario, one with minimal impact to performance, probably less than having _Py_TYPE, etc, exported from python<version> binary.

**Comment by WildCard65** [comment]
@vstinner One way or another, PY_TYPE and co need to have their implementations PURGED from the Python header files for a truly opaque PyObject.

Another thing that needs to be purged are objects that use ob_base in their objects like PyTupleObject, PyType Object, PyListObject to name a few.

I'm willing to replace slib with a precompiled header trio, but that's a can of worms on its own (at least on MSVC).

I'm not budging in regards to "tp_obj_offset" and "tp_obj_size", see above mentioned purging.

**Pull request closed without merging**